[Tran Khoi Nguyen] iP

TrKNguyen opened the pull request:
# NguyenPro  
> “Your mind is for having ideas, not holding them.” – David Allen  

NguyenPro helps you manage your work efficiently so you can focus on what truly matters. It’s:

- **Text-based**  
- **Easy to learn**  
- **FAST & SUPER FAST** to use  

All you need to do is:

1. Clone it from [GitHub Link](https://github.com/TrKNguyen/ip.git).  
2. Run it using Gradle.  
3. Add your tasks.  
4. Let it manage your tasks for you 😉  

And the best part? **It’s FREE!**

## Features:  
- ✅ Managing tasks  
- ⏳ Managing deadlines, event and todo list
- 🔔 Reminders *(coming soon)*  

If you're a Java programmer, you can use it to practice Java too. Here’s the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```
Stay productive with NguyenPro! 🚀
Have fun and stay organized! 😎
Feel free to adjust it to your style!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

TrKNguyen committed:
add: ingore .class file and change name of chatbot

TrKNguyen committed:
Add Level 0. Rename, Greet, Exit

TrKNguyen committed:
Add Level 1. Echo

TrKNguyen committed:
Add Level 2.Add, List

TrKNguyen committed:
Add Level 3. Mark as Done

TrKNguyen committed:
Add Level 3. Mark as Done

TrKNguyen committed:
Add Level 4. ToDos, Events, Deadlines

TrKNguyen committed:
Add A-TextUiTesting

TrKNguyen committed:
Add Level 5. Handle Errors

TrKNguyen committed:
Adding Level 5. Handle Errors

TrKNguyen committed:
Adding Level 6. Delete

TrKNguyen committed:
Adding A-Enums

TrKNguyen committed:
Add comment in the code

TrKNguyen committed:
Add Level 7. Save

TrKNguyen committed:
Add Level 8

TrKNguyen committed:
Add more OOP without testing

TrKNguyen committed:
Add A-MoreOOP

TrKNguyen committed:
Add A-Packages

TrKNguyen committed:
Merge branch 'add-gradle-support'

TrKNguyen committed:
Add A-JUnit

TrKNguyen committed:
Generate nguyen.jar

TrKNguyen committed:
Add A-JavaDoc

TrKNguyen committed:
A-CodingStandard

TrKNguyen committed:
Add Level-9

TrKNguyen committed:
Merge branch 'branch-Level-9'

TrKNguyen committed:
Add A-CheckStyle

TrKNguyen committed:
Add Level 10. GUI

TrKNguyen committed:
Add A-Varargs

TrKNguyen committed:
Modify the Readme.md

TrKNguyen committed:
Add Level-10

vikeedough commented on src/main/java/nguyen/Command.java:
```diff
@@ -0,0 +1,116 @@
+            } else if (type == TaskType.DEADLINE) {
+                int indexBy = item.indexOf("/by");
+                if (item.length() == 8) {
+                    throw new NguyenException("Empty task is valid task");
+                }
+                if (indexBy == -1) {
+                    throw new NguyenException("Dont have deadline");
+                }
+                String task = item.substring(8, indexBy);
+                String by = item.substring(indexBy + 4);
+                taskList.add(new Deadline(task, by));
+            } else if (type == TaskType.EVENT) {
+                int indexFrom = item.indexOf("/from");
+                int indexTo = item.indexOf("/to");
+                if (item.length() == 5) {
+                    throw new NguyenException("Empty task is valid task");
+                }
+                if (indexFrom == -1) {
+                    throw new NguyenException("Dont have starting time");
+                }
+                if (indexTo == -1) {
+                    throw new NguyenException("Dont have ending time");
+                }
+                if (indexFrom > indexTo) {
+                    throw new NguyenException("Wrong syntax order: /from should stand before /to");
+                }
+                String task = item.substring(5, indexFrom);
+                String from = item.substring(indexFrom + 6, indexTo);
+                String to = item.substring(indexTo + 4);
+                taskList.add(new Event(task, from, to));
+            } else {
+                throw new NguyenException("Invalid operation");
+            }
+
+            System.out.println("Got it. I've added this task:");
+            System.out.println(taskList.get(taskList.size() - 1));
+            System.out.println("Now you have " + taskList.size() + " tasks in the list.");
+        }
+        storage.saveTask(taskList);
+    }
```
Maybe consider using a switch statement here? There is a lot of nesting happening with the if-else statements. You can also consider writing separate functions for each case to make what is happening in the code easier to understand!

vikeedough commented on src/main/java/nguyen/DateParser.java:
```diff
@@ -0,0 +1,54 @@
+ */
+public class DateParser {
+    // List of supported date patterns in various formats
+    private static final List<String> DATE_PATTERNS = Arrays.asList(
+            "yyyy-MM-dd HHmm",
+            "yyyy/MM/dd HHmm",
+            "dd/MM/yyyy HHmm",
+            "MM-dd-yyyy HHmm",
+            "yyyy-MM-dd",
+            "yyyy/MM/dd",
+            "dd/MM/yyyy",
+            "MM-dd-yyyy",
+            "dd MMM yyyy",
+            "d MMM yyyy",
+            "MMM d yyyy",
+            "MMM dd yyyy",
+            "MMM dd, yyyy"
+    );
```
I like that you made your code able to handle all different kinds of formats! Well done!

vikeedough commented on src/main/java/nguyen/Parser.java:
```diff
@@ -0,0 +1,43 @@
+ * Parses user input into commands and tasks.
+ */
+public class Parser {
+    /**
+     * Converts a full command string into a Command object.
+     *
+     * @param fullCommand the full user input string
+     * @return a Command object representing the parsed command
+     * @throws NguyenException if the command is invalid
+     */
```
I noticed that there are some JavaDoc comments with no punctuation behind parameter descriptions! May want to rectify to follow the recommended Java coding standard for the course.

vikeedough commented on src/test/java/nguyen/DateParserTest.java:
```diff
@@ -0,0 +1,36 @@
+
+class DateParserTest {
+
+    @Test
+    void testValidDates() {
+        assertEquals(LocalDate.of(2024, 1, 1), DateParser.parseDate("2024-01-01"));
+        assertEquals(LocalDate.of(2024, 1, 1), DateParser.parseDate("2024/01/01"));
+        assertEquals(LocalDate.of(2024, 1, 1), DateParser.parseDate("01/01/2024"));
+        assertEquals(LocalDate.of(2024, 1, 1), DateParser.parseDate("01-01-2024"));
+        assertEquals(LocalDate.of(2024, 2, 1), DateParser.parseDate("1 Feb 2024"));
+        assertEquals(LocalDate.of(2024, 2, 1), DateParser.parseDate("Feb 1 2024"));
+        assertEquals(LocalDate.of(2024, 2, 1), DateParser.parseDate("Feb 01 2024"));
+        assertEquals(LocalDate.of(2024, 2, 1), DateParser.parseDate("Feb 01, 2024"));
+        assertEquals(LocalDate.of(2024, 3, 1), DateParser.parseDate("2024-03-01"));
+        assertEquals(LocalDate.of(2024, 3, 1), DateParser.parseDate("2024/03/01"));
+        assertEquals(LocalDate.of(2024, 3, 1), DateParser.parseDate("01/03/2024"));
+        assertEquals(LocalDate.of(2024, 3, 1), DateParser.parseDate("03-01-2024"));
+    }
```
Well-written test to check all the possible formats!

vikeedough reviewed (commented):
Great job on your iP so far!

aaronlim02 commented on src/main/java/nguyen/Nguyen.java:
```diff
@@ -0,0 +1,85 @@
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+enum TaskType {
+    TODO, DEADLINE, EVENT;
+}
```
I suggest adding a common prefix to each of these constants. 
Doing this indicates that they belong together and makes them appear together when sorted alphabetically.
```suggestion
enum TaskType {
    TASKTYPE_TODO, TASKTYPE_DEADLINE, TASKTYPE_EVENT;
}

aaronlim02 commented on src/main/java/nguyen/Event.java:
```diff
@@ -0,0 +1,34 @@
+     * @return A formatted string representing the event task.
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + from.format(DateTimeFormatter.ofPattern("MMM d yyyy"))
+                + " to: " + to.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")";
+    }
```
I like how you used line wrapping appropriately and avoided the common mistake of only indenting 4 spaces. Good job!

aaronlim02 commented on src/main/java/nguyen/Storage.java:
```diff
@@ -0,0 +1,67 @@
+     * @param taskList The TaskList containing tasks to be saved.
+     */
+    public void saveTask(TaskList taskList) {
+        try {
+            // Clear the existing file content
+            new FileWriter(filePath, false).close();
+            try (BufferedWriter writer = new BufferedWriter(new FileWriter(filePath))) {
+                for (int i = 0; i < taskList.size(); i++) {
+                    writer.write(taskList.get(i).toString());
+                    writer.newLine();
+                }
+            }
```
I think the first try block is redundant and can be removed. You can instead open the BufferedWriter with false for append mode, which should already overwrite the file
```suggestion
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filePath, false))) {
            for (int i = 0; i < taskList.size(); i++) {
                writer.write(taskList.get(i).toString());
                writer.newLine();
            }
        }

aaronlim02 commented on src/test/java/nguyen/DateParserTest.java:
```diff
@@ -0,0 +1,36 @@
+        assertEquals(LocalDate.of(2024, 3, 1), DateParser.parseDate("03-01-2024"));
+    }
+
+    @Test
+    void testInvalidDates() {
+        assertNull(DateParser.parseDate("2024-13-01"));
+        assertNull(DateParser.parseDate("Hello World"));
+        assertNull(DateParser.parseDate(""));
+        assertNull(DateParser.parseDate("2024.02.01"));
+        assertNull(DateParser.parseDate("2/31"));
+    }
+}
```
I like the fact that you considered invalid dates as well!

aaronlim02 reviewed (commented):
Overall well done!

ChauuuLe commented on src/main/java/nguyen/Event.java:
Maybe should wrap Event, Deadline and Todo in a package

ChauuuLe commented on src/main/java/nguyen/Task.java:
Maybe should make Task, TaskList in smaller package

ChauuuLe reviewed (commented):
Good coding standard but try to make smaller package for some files to make the structure more clear

ChauuuLe commented on src/main/java/nguyen/Main.java:
Clear code and easy to understand

ChauuuLe commented on src/main/java/nguyen/Event.java:
Maybe wrap Event, ToDo and Deadline into a smaller package

ChauuuLe commented on src/main/java/nguyen/TaskList.java:
Wrap into Task package

ChauuuLe commented on src/main/java/nguyen/Nguyen.java:
Nice greeting message

ChauuuLe commented on src/main/java/nguyen/Command.java:
```diff
@@ -0,0 +1,116 @@
+            isExit = true;
+            return;
+        }
+        // Displays all tasks in the list
```
Could make subclass for each command to apply polymorphism

TrKNguyen committed:
Add Generate a new JAR file
Generate a new JAR file using Gradle, bundling the JavaFX third-party library into the jarfile. I refer to the Gradle tutorial @SE-EDU/guides to find how.

TrKNguyen committed:
Enhance code quality: Avoid deep nesting
I create a new method 'handleTask' in TaskList to handle task, replace all old code in Command by calling this handleTask function

TrKNguyen committed:
Add the JavaDoc for new handleTask method in TaskList

TrKNguyen committed:
Add the comments to empty catch block.
It is for explalining why the catch block is left empty. Therefore, using emptu catch block is unavoidable

TrKNguyen committed:
Delete deadcode.
There is part of handling redundant errors what have been handled by previous statements within the function

TrKNguyen committed:
Add A-Assertions

TrKNguyen committed:
Add A-Assertions

TrKNguyen committed:
Fix several minor bugs

TrKNguyen committed:
Add A-Streams.
Convert normal printList function to printList using stream

TrKNguyen committed:
Merge pull request #1 from TrKNguyen/branch-A-Assertions

Branch a assertions

TrKNguyen committed:
Merge branch 'master' into branch-A-CodeQuality

TrKNguyen committed:
Merge pull request #3 from TrKNguyen/branch-A-CodeQuality

Branch a code quality

TrKNguyen committed:
Merge branch 'master' into branch-A-Streams

TrKNguyen committed:
Merge pull request #4 from TrKNguyen/branch-A-Streams

Add A-Streams

TrKNguyen committed:
Add A-Ci.
GitHub Actions is CI/CD tool integrated into GitHub.

TrKNguyen committed:
Merge branch 'master' of https://github.com/TrKNguyen/ip

TrKNguyen committed:
Add C-Sort.
Introduce a new feature sort, sort <taskType>. Chatbot will return the list of sorted list by suitable time information of each task

TrKNguyen committed:
Fix A-Streams

TrKNguyen committed:
Add A-Streams

TrKNguyen committed:
Add AI-assisted code for header

TrKNguyen committed:
Add A-BetterGui

TrKNguyen committed:
Add A-MoreErrorHandling

TrKNguyen committed:
Fix save and load task errors in Storage class

TrKNguyen committed:
Update test cases

TrKNguyen committed:
Add A-UserGuide

TrKNguyen committed:
Resolve Gradle Checkstyle violations

TrKNguyen committed:
Add A-Release

TrKNguyen committed:
Add A-MoreTesting

TrKNguyen committed:
Add A-Personality

TrKNguyen committed:
Add A-Release
